src/Sarif.WorkItemFiling/WorkItemFiler.cs

```diff
@@ -35,9 +37,14 @@ public class WorkItemFiler
         /// An object that represents the system (for example, GitHub or Azure DevOps)
         /// to which the work items will be filed.
         /// </param>
-        public WorkItemFiler(FilingTarget filingTarget, IGroupingStrategy groupingStrategy, IFileSystem fileSystem)
+        public WorkItemFiler(
+            FilingTarget filingTarget,
+            IFilteringStrategy filteringStrategy,
+            IGroupingStrategy groupingStrategy,
+            IFileSystem fileSystem)
         {
             _filingTarget = filingTarget ?? throw new ArgumentNullException(nameof(filingTarget));
+            _filteringStrategy = filteringStrategy ?? throw new ArgumentNullException(nameof(filteringStrategy));
```

Comment:
>ArgumentNullException [](start = 64, length = 21)

why aren't we providing happy path values here? I suggest we consider it.

Reply:
I provided a default file system.

There is no obvious default for filtering strategy. If we use "all results", it surprises people who use result matching and don't want to see issues repeated. If we use "new results", it surprises people who don't use result matching and wonder why nothing got filed.

We could define `NewOrUnmarkedFilteringStrategy` that takes either `BaselineState.New` or `BaselineState.None`, because the spec requires that `result.baselineState` be "all or nothing" (and our result matchers implement it this way).

For grouping strategy, `OneResultPerWorkItemStrategy` is probably the least surprising, but it can have serious consequences (filing 100,000 bugs is no joke), so IMO the caller should have to make the choice explicitly.

So I think we should leave the other parameters without defaults for now.

---
In reply to: [306116112](https://github.com/microsoft/sarif-sdk/pull/1582#discussion_r306116112) [](ancestors = 306116112)

Reply:
All results is a fine filtering strategy. After all, what's wrong with 'please assume all these results are relevant to the operation' as a default? The most nuanced default filtering strategy would look like so: examine the first result, if baselineState == None, then we have no baseline information in the log file and so, please pull in all results. If the first result is not None, then all results shall be decorated with a baseline state, in which case we should pull in any result that is either Updated or New. For updated items, the work item thingy will find the existing bug and update its contents, for New items we will file a new item. I suggest you implement this strategy and make it the default. For now, it's fine for Updated bugs to be a no-op.

---
In reply to: [306346682](https://github.com/microsoft/sarif-sdk/pull/1582#discussion_r306346682) [](ancestors = 306346682,306116112)